deltacat/compute/compactor_v2/utils/compaction_utils.py

```diff
@@ -0,0 +1,713 @@
+            for entry in delta.manifest.entries
+        ]
+    )
+    local_merge_options = local_merge_resource_options_provider(
+        estimated_da_size=estimated_da_bytes,
+        estimated_num_rows=estimated_num_records,
+        total_memory_buffer_percentage=params.total_memory_buffer_percentage,
+        round_completion_info=round_completion_info,
+        compacted_delta_manifest=previous_compacted_delta_manifest,
+        ray_custom_resources=params.ray_custom_resources,
+        primary_keys=params.primary_keys,
+        memory_logs_enabled=params.memory_logs_enabled,
+    )
+    local_merge_result = ray.get(
+        mg.merge.options(**local_merge_options).remote(local_merge_input)
+    )
+    total_input_records_count += local_merge_result.input_record_count
+    merge_results = [local_merge_result]
+    return merge_results, total_input_records_count
+
+
+def _hash_bucket(
+    params: CompactPartitionParams,
+    uniform_deltas,
+    hb_options_provider,
+    task_max_parallelism,
+):
+    def hash_bucket_input_provider(index, item) -> dict[str, HashBucketInput]:
+        return {
+            "input": HashBucketInput.of(
+                item,
+                primary_keys=params.primary_keys,
+                hb_task_index=index,
+                num_hash_buckets=params.hash_bucket_count,
+                num_hash_groups=params.hash_group_count,
+                enable_profiler=params.enable_profiler,
+                metrics_config=params.metrics_config,
+                read_kwargs_provider=params.read_kwargs_provider,
+                object_store=params.object_store,
+                deltacat_storage=params.deltacat_storage,
+                deltacat_storage_kwargs=params.deltacat_storage_kwargs,
+                memory_logs_enabled=params.memory_logs_enabled,
+            )
+        }
+
+    hb_tasks_pending = invoke_parallel(
+        items=uniform_deltas,
+        ray_task=hb.hash_bucket,
+        max_parallelism=task_max_parallelism,
+        options_provider=hb_options_provider,
+        kwargs_provider=hash_bucket_input_provider,
+    )
+
+    hb_invoke_end = time.monotonic()
+
+    logger.info(f"Getting {len(hb_tasks_pending)} hash bucket results...")
+    hb_results: List[HashBucketResult] = ray.get(hb_tasks_pending)
+    logger.info(f"Got {len(hb_results)} hash bucket results.")
+
+    return (hb_results, hb_invoke_end)
+
+
+def _merge(
+    params: CompactPartitionParams,
+    task_resource_options_provider,
+    merge_resource_options_provider,
+    all_hash_group_idx_to_size_bytes,
+    all_hash_group_idx_to_num_rows,
+    round_completion_info,
+    previous_compacted_delta_manifest,
+    all_hash_group_idx_to_obj_id,
+    compacted_partition,
+    delete_strategy,
+    delete_file_envelopes,
+    task_max_parallelism,
+) -> tuple[List[MergeResult], float]:
+    merge_options_provider = functools.partial(
+        task_resource_options_provider,
+        pg_config=params.pg_config,
```

Comment:
This is the correct placement. You have merge options provider inside merge. Follow the same pattern for hash bucket.